onnxoptimizer/passes/eliminate_slice_shape.h

```diff
@@ -0,0 +1,136 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+// ATTENTION: The code in this file is highly EXPERIMENTAL.
+// Adventurous users should note that the APIs will probably change.
+
+#pragma once
+
+#include "onnx/defs/tensor_util.h"
+#include "onnxoptimizer/pass.h"
+
+namespace ONNX_NAMESPACE {
+namespace optimization {
+
+struct EliminateSliceShape final : public PredicateBasedPass {
+  explicit EliminateSliceShape()
+      : PredicateBasedPass(PassType::Nop, PassEfficiency::Complete,
+                           PassOptimizationType::Compute) {}
+
+  std::string getPassName() const override {
+    return "eliminate_slice_shape";
+  }
+
+  bool patternMatchPredicate(Node *node) override {
+    return node->kind() == kSlice &&
+           node->inputs()[0]->node()->kind() == Symbol("Shape") &&
+           node->inputs()[0]->node()->input()->has_sizes();
+  }
+
+  bool runTransform(Node *node, Graph &graph,
+                    NodeDestroyType &destroy_current) override {
+    Node *shape_node = node->inputs()[0]->node();
+    const auto &dims_of_shape_node_input = shape_node->input()->sizes();
+    std::vector<Dimension> result_of_shape_op;
+
+    auto add_y_if_negative = [](int64_t x, int64_t y) -> int64_t {
+      return x < 0 ? x + y : x;
+    };
+    {
+      const int64_t shape_start =
+          add_y_if_negative(shape_node->hasAttribute(Symbol("start"))
+                                ? shape_node->i(Symbol("start"))
+                                : 0,
+                            dims_of_shape_node_input.size());
+      const int64_t shape_end =
+          add_y_if_negative(shape_node->hasAttribute(Symbol("end"))
+                                ? shape_node->i(Symbol("end"))
+                                : dims_of_shape_node_input.size(),
+                            dims_of_shape_node_input.size());
+
+      for (int i = shape_start; i < shape_end; ++i) {
+        result_of_shape_op.push_back(dims_of_shape_node_input[i]);
+      }
+    }
+    auto fetch_first_value_of_tensor = [&graph](const Value *v,
+                                                int64_t &i_d) -> bool {
+      const auto &initializer_names = graph.initializer_names();
+      const uint32_t kind = v->node()->kind();
+      const Tensor *tensor = nullptr;
+      if (kind == kConstant) {
+        tensor = &v->node()->t(kvalue);
+      } else if (kind == kParam &&
+                 std::find(initializer_names.cbegin(), initializer_names.cend(),
+                           v->uniqueName()) != initializer_names.cend()) {
```

Comment:
remove line 58 and replace these lines with

```suggestion
      } else if (graph.is_constant_initializer(v)) {
```

`graph.is_constant_initializer` is added in https://github.com/onnx/optimizer/pull/83